parquet/src/main/java/org/apache/iceberg/parquet/vectorized/VectorizedParquetReader.java

```diff
@@ -0,0 +1,290 @@
+      this.file = toCopy.file;
+      this.options = toCopy.options;
+      this.projection = toCopy.projection;
+      this.model = toCopy.model;
+      this.rowGroups = toCopy.rowGroups;
+      this.shouldSkip = toCopy.shouldSkip;
+      this.totalValues = toCopy.totalValues;
+      this.reuseContainers = toCopy.reuseContainers;
+      this.batchSize = toCopy.batchSize;
+    }
+
+    ParquetFileReader reader() {
+      if (reader != null) {
+        reader.setRequestedSchema(projection);
+        return reader;
+      }
+
+      ParquetFileReader newReader = newReader(file, options);
+      newReader.setRequestedSchema(projection);
+      return newReader;
+    }
+
+    VectorizedReader model() {
+      return model;
+    }
+
+    boolean[] shouldSkip() {
+      return shouldSkip;
+    }
+
+    long totalValues() {
+      return totalValues;
+    }
+
+    boolean reuseContainers() {
+      return reuseContainers;
+    }
+
+    int batchSize() {
+      return batchSize;
+    }
+
+    private static ParquetFileReader newReader(InputFile file, ParquetReadOptions options) {
+      try {
+        return ParquetFileReader.open(ParquetIO.file(file), options);
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open Parquet file: %s", file.location());
+      }
+    }
+
+    ReadConf<T> copy() {
+      return new ReadConf<>(this);
+    }
+  }
+
+  private ReadConf conf = null;
+
+  private ReadConf init() {
+    if (conf == null) {
+      ReadConf readConf = new ReadConf(
+          input, options, expectedSchema, filter, batchReaderFunc, reuseContainers, caseSensitive, batchSize);
+      this.conf = readConf.copy();
+      return readConf;
+    }
+
+    return conf;
+  }
+
+  @Override
+  public Iterator iterator() {
+    FileIterator iter = new FileIterator(init());
+    addCloseable(iter);
+    return iter;
+  }
+
+  private static class FileIterator<T> implements Iterator<T>, Closeable {
+    private final ParquetFileReader reader;
+    private final boolean[] shouldSkip;
+    private final VectorizedReader<T> model;
+    private final long totalValues;
+    private final int batchSize;
+
+    private int nextRowGroup = 0;
+    private long nextRowGroupStart = 0;
+    private long valuesRead = 0;
+    private T last = null;
+
+    FileIterator(ReadConf conf) {
+      this.reader = conf.reader();
+      this.shouldSkip = conf.shouldSkip();
+      this.model = conf.model();
+      this.totalValues = conf.totalValues();
+      this.model.reuseContainers(conf.reuseContainers());
+      this.batchSize = conf.batchSize();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return valuesRead < totalValues;
+    }
+
+    @Override
+    public T next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      if (valuesRead >= nextRowGroupStart) {
+        advance();
+      }
+      this.last = model.read();
+      valuesRead += Math.min(nextRowGroupStart - valuesRead, batchSize);
```

Comment:
This assumes that the number of rows read matches the number of rows expected, instead of using the actual number of rows read or requesting the model read some number of rows. I think this should probably pass in the required row count to ensure the accounting doesn't diverge.

Reply:
Made the change.